net/cgi-io/src/main.c

```diff
@@ -640,14 +641,18 @@ main_upload(int argc, char *argv[])
 	{
 		rem = multipart_parser_execute(p, buf, len);
 
-		if (rem < len)
+		if (rem < len) {
+			len = rem;
 			break;
+		}
 	}
 
 	multipart_parser_free(p);
-
 	/* read remaining post data */
-	while ((len = read(0, buf, sizeof(buf))) > 0);
+
+	while (len == BLOCK_SIZE) {
+		len = read(0, buf, sizeof(buf));
+	};
```

Comment:
I think the second while loop should be removed entirely. Instead add a boolean option `done` which is set to `true` when `rem < len`. Then only execute `multipart_parser_execute()` when `done` is false.

Example:

```
int rem, len;
bool done = false;
...

while ((len = read(0, buf, sizeof(buf))) > 0)
{
    if (!done) {
        rem = multipart_parser_execute(p, buf, len);
        done = (rem < len);
    }
}
 
multipart_parser_free(p);
```

Reply:
I added a break condition if the rem is less than a block
Without this the program locks as read() try to get another block from stdin

Reply:
I think this is not the right abort condition. The read() will also return if STDIN is closed. For uwsgi there is an option for that:
`cgi-close-stdin-on-eof = true`